test/library/packages/Itertools/Itertools.chpl

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright 2004-2019 Cray Inc.
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+  This module serves as a toolkit of common serial and parallel
+  iterators. This is currently being worked upon, and contains
+  only the ``repeat`` iterator tool as of now.
+*/
+
+module Itertools {
+
+  /*
+    Returns an object over and over again, a specified
+    number of times.
+
+
+    :arg arg: The object to be returned
+    :type arg: `?`
+
+    :arg times: The number of times to return
+    :type times: `int`
+
+    :yields: Object in the range ``1..times``
+
+    :throws: ``IllegalArgumentError`` on parallel infinite iteration (see below)
+
+
+    If the argument ``times`` has the value 0, it will return
+    the object an infinite number of times.
+
+    This iterator can be called in serial and parallel zippered contexts.
+
+    .. note::
+      This iterator is not suitable for parallel infinite iteration i.e.
+      avoid using zippered, ``forall``, or ``coforall`` loops with the
+      ``times`` argument set to 0.
+  */
+
+  // Serial iterator
+
+  iter repeat (arg, times = 0) {
+    if times == 0 then
+      for 1.. do yield arg;
+    else
+      for 1..#times do yield arg;
+  }
+
+  // Standalone parallel iterator
+
+  pragma "no doc"
+  iter repeat (param tag: iterKind, arg, times = 0) throws
+      where tag == iterKind.standalone {
+    if times == 0 then
+      throw new owned IllegalArgumentError(
+          "Infinite iteration not supported for parallel loops.");
+    else
+      forall 1..#times do yield arg;
+  }
+
+  use RangeChunk;
+
+  // Parallel iterator - Leader
+
+  pragma "no doc"
+  iter repeat (param tag: iterKind, arg, times = 0) throws
+      where tag == iterKind.leader {
+
+    const numTasks = if dataParTasksPerLocale > 0 then dataParTasksPerLocale
+                                                  else here.maxTaskPar;
+
+    if times == 0 then
+      throw new owned IllegalArgumentError(
+          "Infinite iteration not supported for parallel loops.");
+    else
+      coforall tid in 0..#numTasks {
+        const working_iters = chunk(0..#times, numTasks, tid);
```

Comment:
Just ran testing on a 48 core machine and got a runtime error here, because `numTasks > times`.

I think you need to add some special case logic, like:

```chpl
if numTasks > times then numTasks = times;
```

You could use a different variable since `numTasks` is const or make `numTasks` var.

Reply:
Woah, I never thought of this. One learns new things everyday :)

Fixed it by making `numTasks` variable and adding the line mentioned above.

Reply:
@ben-albrecht, shouldn't there also be some kind of error/halt produced when such a case is encountered by the built-in `chunk()` function?